ikarus/finiteelements/mechanics/kirchhoffloveshell.hh

```diff
@@ -180,30 +203,179 @@ namespace Ikarus {
         volumeLoad;
-    std::function<Eigen::Vector<double, Traits::worlddim>(const Eigen::Vector<double, Traits::worlddim>&,
-                                                          const double&)>
+    std::function<Eigen::Vector<double, Traits::worlddim>(const Dune::FieldVector<double, Traits::worlddim> &,
+                                                          const double &)>
         neumannBoundaryLoad;
-    const BoundaryPatch<GridView>* neumannBoundary;
+    const BoundaryPatch<GridView> *neumannBoundary;
     mutable Dune::BlockVector<Dune::RealTuple<double, Traits::dimension>> dispAtNodes;
+    DefaultMembraneStrain membraneStrain;
     double emod_;
     double nu_;
     double thickness_;
     size_t numberOfNodes{0};
     int order{};
 
   protected:
+    auto computeMaterialAndStrains(const Dune::FieldVector<double, 2> &gpPos, int gpIndex, const Geometry &geo,
+                                   const auto &uFunction) const {
+      using ScalarType = typename std::remove_cvref_t<decltype(uFunction)>::ctype;
+      using namespace Dune;
+      using namespace Dune::DerivativeDirections;
+      const auto [X, Jd, Hd]              = geo.impl().zeroFirstAndSecondDerivativeOfPosition(gpPos);
+      const auto J                        = toEigen(Jd);
+      const auto H                        = toEigen(Hd);
+      const Eigen::Matrix<double, 2, 2> A = J * J.transpose();
+      Eigen::Matrix<double, 3, 3> G;
+      G.setZero();
+      G.block<2, 2>(0, 0)                    = A;
+      G(2, 2)                                = 1;
+      const Eigen::Matrix<double, 3, 3> GInv = G.inverse();
+      const auto C                           = materialTangent(GInv);
+
+      Eigen::Vector3<ScalarType> epsV = membraneStrain.value(gpPos, geo, uFunction);
+
+      const auto &Ndd                             = localBasis.evaluateSecondDerivatives(gpIndex);
+      const auto uasMatrix                        = Dune::viewAsEigenMatrixAsDynFixed(uFunction.coefficientsRef());
+      const auto hessianu                         = Ndd.transpose().template cast<ScalarType>() * uasMatrix;
+      const Eigen::Matrix3<ScalarType> h          = H + hessianu;
+      const Eigen::Matrix<ScalarType, 3, 2> gradu = toEigen(uFunction.evaluateDerivative(
+          gpIndex, Dune::wrt(spatialAll), Dune::on(Dune::DerivativeDirections::referenceElement)));
+      const Eigen::Matrix<ScalarType, 2, 3> j     = J + gradu.transpose();
+      const Eigen::Vector3<ScalarType> a3N        = (j.row(0).cross(j.row(1)));
+      const Eigen::Vector3<ScalarType> a3         = a3N.normalized();
+      Eigen::Vector<ScalarType, 3> bV             = h * a3;
+      bV(2) *= 2;  // Voigt notation requires the two here
+      const auto BV     = toVoigt(toEigen(geo.impl().secondFundamentalForm(gpPos)));
+      const auto kappaV = (BV - bV).eval();
+      return std::make_tuple(C, epsV, kappaV, j, J, h, H, a3N, a3);
```

Comment:
I think a small struct would be more readable. As far as I know structural bindings (i.e. `const auto [C, eps, ...] =computeMaterial()` should still work